File: sdks/ruby/test/test_client.rb
# frozen_string_literal: true

require "minitest/autorun"
require "webrick"
require "json"

require_relative "../lib/featuresignals"

class TestClient < Minitest::Test
  CANNED_FLAGS = { "feature-a" => true, "banner" => "hello", "count" => 42 }.freeze

  class << self
    attr_accessor :server, :base_url
  end

  def setup
    return if self.class.server

    self.class.server = WEBrick::HTTPServer.new(
      Port: 0,
      Logger: WEBrick::Log.new("/dev/null"),
      AccessLog: []
    )

    self.class.server.mount_proc "/" do |_req, res|
      res["Content-Type"] = "application/json"
      res.body = JSON.generate(CANNED_FLAGS)
    end

    port = self.class.server.config[:Port]
    self.class.base_url = "http://127.0.0.1:#{port}"
    Thread.new { self.class.server.start }
    sleep 0.1 # allow server to bind
  end

  def make_client(**overrides)
    opts = FeatureSignals::ClientOptions.new(
      env_key: "dev",
      base_url: self.class.base_url,
      polling_interval: 60,
      **overrides
    )
    client = FeatureSignals::Client.new("test-key", opts)
    client.wait_for_ready(timeout: 5)
    client
  end

  def test_bool_variation
    c = make_client
    assert_equal true, c.bool_variation("feature-a", FeatureSignals::EvalContext.new(key: "u1"), false)
  ensure
    c&.close
  end

  def test_string_variation
    c = make_client
    assert_equal "hello", c.string_variation("banner", FeatureSignals::EvalContext.new(key: "u1"), "")
  ensure
    c&.close
  end

  def test_number_variation
    c = make_client
    assert_equal 42, c.number_variation("count", FeatureSignals::EvalContext.new(key: "u1"), 0)
  ensure
    c&.close
  end

  def test_fallback_on_missing_key
    c = make_client
    assert_equal false, c.bool_variation("missing", FeatureSignals::EvalContext.new(key: "u1"), false)
  ensure
    c&.close
  end

  def test_fallback_on_wrong_type
    c = make_client
    assert_equal "nope", c.string_variation("feature-a", FeatureSignals::EvalContext.new(key: "u1"), "nope")
  ensure
    c&.close
  end

  def test_all_flags
    c = make_client
    flags = c.all_flags
    assert_includes flags.keys, "feature-a"
    assert_includes flags.keys, "banner"
    assert_includes flags.keys, "count"
  ensure
    c&.close
  end

  def test_ready
    c = make_client
    assert c.ready?
  ensure
    c&.close
  end

  def test_on_ready_callback
    called = Queue.new
    opts = FeatureSignals::ClientOptions.new(
      env_key: "dev",
      base_url: self.class.base_url,
      polling_interval: 60
    )
    c = FeatureSignals::Client.new("test-key", opts, on_ready: -> { called.push(true) })

    result = begin
               called.pop(true)
             rescue ThreadError
               sleep 1
               begin
                 called.pop(true)
               rescue ThreadError
                 false
               end
             end
    assert result, "on_ready callback should have been called"
  ensure
    c&.close
  end

  Minitest.after_run do
    TestClient.server&.shutdown
  end
end

File: sdks/python/tests/test_client.py
"""Unit tests for the FeatureSignals Python SDK."""

from __future__ import annotations

import json
import threading
from http.server import HTTPServer, BaseHTTPRequestHandler
from unittest import TestCase

from featuresignals import FeatureSignalsClient, ClientOptions, EvalContext


class FlagHandler(BaseHTTPRequestHandler):
    """Minimal HTTP handler that returns canned flag values."""

    flags: dict = {"feature-a": True, "banner": "hello", "count": 42}

    def do_GET(self, *_: object) -> None:
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(self.flags).encode())

    def log_message(self, *_: object) -> None:
        pass


def _start_server() -> tuple[HTTPServer, str]:
    server = HTTPServer(("127.0.0.1", 0), FlagHandler)
    port = server.server_address[1]
    t = threading.Thread(target=server.serve_forever, daemon=True)
    t.start()
    return server, f"http://127.0.0.1:{port}"


class TestClient(TestCase):
    server: HTTPServer
    base_url: str

    @classmethod
    def setUpClass(cls) -> None:
        cls.server, cls.base_url = _start_server()

    @classmethod
    def tearDownClass(cls) -> None:
        cls.server.shutdown()

    def _make_client(self) -> FeatureSignalsClient:
        opts = ClientOptions(env_key="dev", base_url=self.base_url, polling_interval=60.0)
        client = FeatureSignalsClient("test-key", opts)
        client.wait_for_ready(5.0)
        return client

    def test_bool_variation(self) -> None:
        c = self._make_client()
        self.assertTrue(c.bool_variation("feature-a", EvalContext("u1"), False))
        c.close()

    def test_string_variation(self) -> None:
        c = self._make_client()
        self.assertEqual(c.string_variation("banner", EvalContext("u1"), ""), "hello")
        c.close()

    def test_number_variation(self) -> None:
        c = self._make_client()
        self.assertEqual(c.number_variation("count", EvalContext("u1"), 0), 42)
        c.close()

    def test_fallback_on_missing_key(self) -> None:
        c = self._make_client()
        self.assertFalse(c.bool_variation("missing", EvalContext("u1"), False))
        c.close()

    def test_fallback_on_wrong_type(self) -> None:
        c = self._make_client()
        self.assertEqual(c.string_variation("feature-a", EvalContext("u1"), "nope"), "nope")
        c.close()

    def test_all_flags(self) -> None:
        c = self._make_client()
        flags = c.all_flags()
        self.assertIn("feature-a", flags)
        self.assertIn("banner", flags)
        c.close()

    def test_is_ready(self) -> None:
        c = self._make_client()
        self.assertTrue(c.is_ready())
        c.close()

    def test_on_ready_callback(self) -> None:
        called = threading.Event()
        opts = ClientOptions(env_key="dev", base_url=self.base_url, polling_interval=60.0)
        c = FeatureSignalsClient("test-key", opts, on_ready=called.set)
        self.assertTrue(called.wait(5.0))
        c.close()
